### PR TITLE
SDK-604: Sandbox

### PR DESCRIPTION
**Pull request by davidgrayston**
### Added
- `SandboxClientBuilder` builds `SandboxClient`
- `SandboxAttributeBuilder` builds `SandboxAttribute`
- `SandboxAgeVerificationBuilder` builds `SandboxAgeVerification`
- `SandboxAnchorBuilder` builds `SandboxAnchor`
- `TokenRequestBuilder` builds `TokenRequest`
- `YotiDate.fromDateString()`
- `YotiDate.toISODateString()`

### Fixed
- Exception message tests now check messages are correct (regression when switching to Jest)

### CODE REVIEW

File: src/sandbox/client.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        })
+        .catch((err) => {
+          console.log(`Error retrieving requested data: ${err}`);
+          return reject(err);
+        });
+    });
+  }
+}
+
+module.exports = {
```
Comment (MrBurtyyy) [inline src/sandbox/client.js]: Are these better off being exported as `module.exports = SandboxClient`?  Generally I think you only return an object if you are exporting more than one class/component as you are just creating an object and then instantly destructuring it when you import

File: src/sandbox/index.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const { SandboxClientBuilder } = require('./client.builder');
```
Comment (MrBurtyyy) [inline src/sandbox/index.js]: Further to my previous comment, these would become for example: `const SandboxClient = require('./client.builder');`

File: src/sandbox/profile/request/attribute/anchor.builder.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * @class SandboxAnchorBuilder
+ */
+class SandboxAnchorBuilder {
+  /**
+   * Setup initial property values.
+   */
+  constructor() {
+    this.anchors = [];
```
Comment (MrBurtyyy) [inline src/sandbox/profile/request/attribute/anchor.builder.js]: Whats this used for?

File: src/sandbox/profile/request/attribute/anchor.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
+/**
+ * @class SandboxAnchor
+ */
+class SandboxAnchor {
+  /**
+   * @param {string} type
+   * @param {string} value
+   * @param {*} subType
```
Comment (MrBurtyyy) [inline src/sandbox/profile/request/attribute/anchor.js]: Isn't sub-type always a string?

File: src/sandbox/profile/request/attribute/attribute.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+   * @returns {boolean}
+   */
+  getOptional() {
+    return this.optional;
+  }
+
+  /**
+   * @returns {SandboxAnchor[]}
+   */
+  getAnchors() {
```
Comment (MrBurtyyy) [inline src/sandbox/profile/request/attribute/attribute.js]: Does Node not implement the same helper methods as Python and Java to easily filter the sources and verifiers?  Don't think it's needed, but would be good to keep them all similar?

File: src/sandbox/profile/request/token.builder.js [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    return this.withAttribute(sandboxAttribute);
+  }
+
+  /**
+   * @returns {TokenRequest}
+   */
+  build() {
+    return new TokenRequest(
+      this.rememberMeId,
+      Object.keys(this.attributes).map(k => this.attributes[k])
```
Comment (MrBurtyyy) [inline src/sandbox/profile/request/token.builder.js]: Would it not be better to store the attributes in an array originally?  By using a keyed dictionary, aren't we limiting the sandbox token to only allow one attribute with a certain name? (Also we wouldn't have to convert from a dictionary to array when building the request)

File: src/sandbox/client.js (hunk 1 above)
Comment (davidgrayston) [reply src/sandbox/client.js]: Fixed in 6c963d8

File: src/sandbox/index.js (hunk 2 above)
Comment (davidgrayston) [reply src/sandbox/index.js]: Fixed in 6c963d8

File: src/sandbox/profile/request/attribute/anchor.builder.js (hunk 3 above)
Comment (davidgrayston) [reply src/sandbox/profile/request/attribute/anchor.builder.js]: Removed in e47de2c

File: src/sandbox/profile/request/attribute/attribute.js (hunk 5 above)
Comment (davidgrayston) [reply src/sandbox/profile/request/attribute/attribute.js]: _Java SDK_ has only `getAnchors()` - I think this is ok for sandbox testing purposes

File: src/sandbox/profile/request/token.builder.js (hunk 6 above)
Comment (davidgrayston) [reply src/sandbox/profile/request/token.builder.js]: _Java SDK_ uses a map similar to this - as discussed we can change to array/list when we add support for multiple attributes with the same name, but different constraints

File: src/sandbox/profile/request/attribute/anchor.js (hunk 4 above)
Comment (davidgrayston) [reply src/sandbox/profile/request/attribute/anchor.js]: Corrected in e6aa460

**Comment by davidgrayston** [comment]
I've moved sandbox out of `src` in e6aa460...4077687

Sandbox can now be required as follows:

```javascript
const {	
  SandboxClientBuilder,	
  SandboxAttributeBuilder,	
  SandboxAgeVerificationBuilder,	
  SandboxAnchorBuilder,	
  TokenRequestBuilder,	
} = require('yoti/sandbox');
```

**Pull request merged**